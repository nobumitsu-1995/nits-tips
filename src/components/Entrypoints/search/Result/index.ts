export { Container as Result } from './Container'
