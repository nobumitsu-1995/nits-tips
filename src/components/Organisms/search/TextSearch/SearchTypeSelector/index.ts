export {
  SearchTypeSelector,
  SEARCH_TYPE,
  type SearchType,
} from './SearchTypeSelector'
