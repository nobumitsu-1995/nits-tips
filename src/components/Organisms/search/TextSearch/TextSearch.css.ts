import { style } from '@vanilla-extract/css'

export const textSearch = style({})

export const formContainer = style({
  marginTop: 24,
  maxWidth: 570,
  display: 'flex',
  flexDirection: 'column',
  gap: 12,
})
