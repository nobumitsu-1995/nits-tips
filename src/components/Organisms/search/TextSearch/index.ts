export { TextSearch } from './TextSearch'
export { SEARCH_TYPE, type SearchType } from './SearchTypeSelector'
