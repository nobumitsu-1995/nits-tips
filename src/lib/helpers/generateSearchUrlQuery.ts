import {
  SEARCH_TYPE,
  type SearchType,
} from '@/components/Organisms/search/TextSearch'

type FilterSearch = {
  tags: string[]
  category: string
}

type WordSearch = {
  searchType: SearchType
  searchWord: string
}

type Args = FilterSearch | WordSearch

const isWordSearch = (args: Args): args is WordSearch =>
  'searchType' in args &&
  Object.values(SEARCH_TYPE).includes(args.searchType) &&
  'searchWord' in args
const isFilterSearch = (args: Args): args is WordSearch =>
  'category' in args && 'tags' in args

const generateWordSearchUrl = ({ searchWord, searchType }: WordSearch) =>
  `searchWord=${searchWord}&searchType=${SEARCH_TYPE[searchType]}`
const generateFilterSearchUrl = ({ tags, category }: FilterSearch) =>
  tags.length > 0
    ? `category=${category}&tag=${tags.join(',')}`
    : `category=${category}`

export const generateSearchUrlQuery = (args: Args): string => {
  if (isWordSearch(args)) {
    return generateWordSearchUrl(args)
  }
  if (isFilterSearch(args)) {
    return generateFilterSearchUrl(args)
  }

  return ''
}
