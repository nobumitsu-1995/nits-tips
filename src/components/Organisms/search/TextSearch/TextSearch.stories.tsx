import type { Meta, StoryObj } from '@storybook/react'
import { TextSearch } from './TextSearch'

type ComponentType = typeof TextSearch
type Story = StoryObj<ComponentType>

export default {
  component: TextSearch,
} as Meta<ComponentType>

export const Default: Story = {}
