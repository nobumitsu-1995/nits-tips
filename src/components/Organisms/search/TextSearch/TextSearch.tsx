import React, { useId, useState } from 'react'
import { Heading } from '@Atoms/Heading'
import * as styles from './TextSearch.css'
import { SearchForm } from '../SearchForm'
import {
  SearchTypeSelector,
  SEARCH_TYPE,
  type SearchType,
} from './SearchTypeSelector'

export const TextSearch: React.FC = () => {
  const [searchType, setSearchType] = useState<SearchType>(SEARCH_TYPE.all)
  const [searchWord, setSearchWord] = useState('')
  const a11yId = useId()
  const handleChangeSearch = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearchWord(e.target.value)
  }
  return (
    <section>
      <Heading id={a11yId} text="ワード検索" />
      <div className={styles.formContainer}>
        <SearchTypeSelector
          searchType={searchType}
          setSearchType={setSearchType}
        />
        <SearchForm
          id={a11yId}
          searchType={searchType}
          searchWord={searchWord}
          handleChangeSearch={handleChangeSearch}
        />
      </div>
    </section>
  )
}
