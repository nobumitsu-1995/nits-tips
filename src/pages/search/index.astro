---
import { FilterSearch } from '@/components/Organisms/search/FilterSearch'
import { TextSearch } from '@/components/Organisms/search/TextSearch'
import { Border } from '@/components/UI/Atoms/Border'
import { Card } from '@/components/UI/Atoms/Card'
import BaseLayout from '@layout/BaseLayout.astro'
import { getAllCategories } from '@/lib/API/microCMS/getAllCategories'
import { getAllTags } from '@/lib/API/microCMS/getAllTags'
import * as styles from '@/styles/pages/search.css'

export const prerender = false

const pageTitle = 'nits-tips | 検索画面'
const description = 'nobumitsu-1995による技術tipsの備忘録集。検索画面です。'
const categories = await getAllCategories()
const tags = await getAllTags()
---

<BaseLayout pageTitle={pageTitle} description={description}>
  <main class={styles.container}>
    <div class={styles.cardContainer}>
      <Card padding='48px 28px'>
        <TextSearch client:load />
        <Border />
        <FilterSearch client:load categoriesData={categories} tagsData={tags} />
      </Card>
    </div>
  </main>
</BaseLayout>
