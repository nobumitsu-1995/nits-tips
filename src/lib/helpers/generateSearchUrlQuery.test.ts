import { SEARCH_TYPE } from '@/components/Organisms/search/TextSearch'
import { generateSearchUrlQuery } from './generateSearchUrlQuery'

const FILTER_SEARCH_PARAMS = {
  tags: ['testTag1', 'testTag2'],
  category: 'testCategory',
}

const WORD_SEARCH_PARAMS = {
  searchType: SEARCH_TYPE.all,
  searchWord: 'testSearchWord',
}

test('絞り込み検索の引数が渡された時、絞込検索用のURLクエリを作成する', () => {
  expect(generateSearchUrlQuery(FILTER_SEARCH_PARAMS)).toMatchSnapshot()
})

test('絞り込み検索の引数でtagsが空配列で渡された時、絞込検索用のURLクエリを作成する', () => {
  expect(
    generateSearchUrlQuery({ ...FILTER_SEARCH_PARAMS, tags: [] }),
  ).toMatchSnapshot()
})

test('ワード検索の引数が渡された時、ワード検索用のURLクエリを作成する', () => {
  expect(generateSearchUrlQuery(WORD_SEARCH_PARAMS)).toMatchSnapshot()
})

test('ワード検索の引数でスペース有のキーワードが渡された時、ワード検索用のURLクエリを作成する', () => {
  expect(
    generateSearchUrlQuery({
      ...WORD_SEARCH_PARAMS,
      searchWord: 'testSearchWord1 testSearchWord2',
    }),
  ).toMatchSnapshot()
})
