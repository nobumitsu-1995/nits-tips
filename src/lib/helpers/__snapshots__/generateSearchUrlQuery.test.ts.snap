// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ワード検索の引数が渡された時、ワード検索用のURLクエリを作成する 1`] = `"searchWord=testSearchWord&searchType=all"`;

exports[`絞り込み検索の引数が渡された時、絞込検索用のURLクエリを作成する 1`] = `"category=testCategory&tag=testTag1,testTag2"`;

exports[`絞り込み検索の引数でtagsが空配列で渡された時、絞込検索用のURLクエリを作成する 1`] = `"category=testCategory"`;
